File: kotlin/Sudoku.kt
import java.io.File


fun main(args: Array<String>) {
    val puzzle = stringToPuzzle(File(args[0]).readText())
    if (puzzle.size != 81) {
        throw IllegalArgumentException("Puzzle size (${puzzle.size}) != 81")
    }
    println(puzzleToString(solve(puzzle)))
}

fun solve(puzzle: List<Byte>): List<Byte> {
    var possible = puzzle.map { if (it == 0.toByte()) (1..9).map { it.toByte() }.toSet() else setOf(it) }
    while (true) {
        val next = possible.map { it.toMutableSet() }
        for (i in (0..80)) {
            if (next[i].size == 1) {
                for (buddy in buddies(i)) {
                    next[buddy].remove(next[i].first())
                }
            }
        }
        if (next == possible) break
        possible = next
    }
    return possible.map { if (it.size == 1) it.first() else 0 }
}

/**
 * Returns a list of indices that are in the same row, column, or box.
 */
fun buddies(i: Int): List<Int> {
    val row = i / 9
    val col = i % 9
    val box = 3 * (row / 3) + col / 3
    return (0..80).filter {
        val row2 = it / 9
        val col2 = it % 9
        val box2 = 3 * (row2 / 3) + col2 / 3
        it != i && (row == row2 || col == col2 || box == box2)
    }
}

fun puzzleToString(puzzle: List<Byte>): String {
    return (0..80 step 9).map { puzzle.slice(it..it + 8).joinToString(" ") }.joinToString("\n")
}

fun stringToPuzzle(v: String): List<Byte> {
    return v.trim().split(regex = "\\s+".toRegex()).map { it.toByte() }
}

File: typescript/sudoku.ts
function stringToPuzzle(s: string): number[] {
    const puzzle = s.match(/\d/g)!.map(Number);
    if (puzzle.length !== 81) {
        throw new Error('Puzzle must be 81 digits long');
    }
    return puzzle;
}

function puzzleToString(puzzle: number[]): string {
    return Array.from({ length: 9 }, (_, i) => puzzle.slice(i * 9, i * 9 + 9).join(' ')).join('\n');
}

function buddies(idx: number): number[] {
    const row = Math.floor(idx / 9);
    const col = idx % 9;
    const box = 3 * Math.floor(row / 3) + Math.floor(col / 3);

    return Array.from({ length: 81 }, (_, idx2) => idx2).filter(idx2 => {
        const row2 = Math.floor(idx2 / 9);
        const col2 = idx2 % 9;
        const box2 = 3 * Math.floor(row2 / 3) + Math.floor(col2 / 3);
        return idx2 !== idx && (row === row2 || col === col2 || box === box2);
    });
}

function solve(puzzle: number[]): number[] {
    let possible = puzzle.map(value => value === 0 ? new Set(Array.from({ length: 9 }, (_, i) => i + 1)) : new Set([value]));

    while (true) {
        const next = possible.map(set => new Set(set));

        for (let i = 0; i < 81; i++) {
            if (next[i].size === 1) {
                const value = Array.from(next[i])[0];
                for (const buddy of buddies(i)) {
                    next[buddy].delete(value);
                }
            }
        }

        if (next.every((set, idx) => set.size === possible[idx].size && Array.from(set).every(value => possible[idx].has(value)))) {
            break;
        }

        possible = next;
    }

    return possible.map(set => set.size === 1 ? set.values().next().value : 0);
}

console.log(puzzleToString(solve(stringToPuzzle(require('fs').readFileSync(0, 'utf-8')))));